MetaMask/metamask-mobile
# refactor: generate nonces using nonce tracker

## **Description**

### Use NonceTracker package

When sending transactions, the `nonce` value is currently generated using an `eth_getTransactionCount` query to determine the next nonce according to the network, plus a caching layer in the `NetworkController` which intercepts this query result to accommodate any pending transactions sent from the wallet.

This PR replaces all of the above and generates the `nonce` value using only the `NonceTracker` package, which internally continues to check the network nonce, and also factor pending transactions but by explicitly checking them, rather than relying on a cached value.

This same package is used in the extension.

Following the previous bug concerning incorrect `nonce` values, the `NonceTracker` usage now explicitly ignores incoming transfer transactions using the `isTransfer` property.

### Synchronise custom nonce logic 

The `NonceTracker` is now also used to propose nonces when `Customize transaction nonce` is enabled, and the related UI logic now utilises the `TransactionController` to ensure this logic remains synchronised in future.

### Generate submit history

Finally, as a means to protect the user following any nonce issues in future, a `submitHistory` array is now automatically populated whenever a transaction is successfully submitted.

This records the following data per transaction in order to support debugging efforts even after `Reset account` has been used:

- Chain ID
- Transaction Hash
- Network Type
- Network URL
- Origin
- Raw Transaction
- Time
- Transaction Parameters

This history is currently limited to the last 100 submitted transactions.

A migration has been added to generate this information from existing wallets using the transaction metadata.

## **Related issues**

Fixes: #7673 

## **Manual testing steps**

Full transaction regression.

Additional focus on all nonce edge cases including:

- Presence of incoming transfer transaction with higher nonces.
- Presence of standard incoming transactions.
- Using `Reset Account` after sending a transaction with a higher nonce.
- Cancelling a transaction with a high nonce then sending a normal transaction.
- Sending multiple pending transactions.
- Switching networks during pending transactions and after confirmed transactions.
- Using custom nonce values.

Verify the correct `submitHistory` is now added to state logs after the following scenarios:

- Use of alternate networks.
- Use of `Reset account` button.
- Upgrade from previous versions using existing transaction state.

## **Screenshots/Recordings**

### **Before**

### **After**

[Pending]

## **Pre-merge author checklist**


## **Pre-merge reviewer checklist**
